src/Compilers/CSharp/Portable/Symbols/TypeSymbolWithAnnotations.cs

```diff
@@ -11,6 +11,23 @@
 
 namespace Microsoft.CodeAnalysis.CSharp.Symbols
 {
+    internal enum NullableAnnotation : byte
+    {
+        Unknown,     // No information. Think oblivious.
+        NotNullable, // For string, int, T
+        Nullable,    // For string?, T? where T : class; and for int?, T? where T : struct.
+        NotNullableBasedOnAnalysis, // Explicitly set by flow analysis
+        NullableBasedOnAnalysis, // Explicitly set by flow analysis
+    }
```

Comment:
If the additional field (`NotNullableBasedOnAnalysis`) is for marking unconstrained type parameter references that are not nullable, perhaps the `Nullable` and `NotNullable` values should be independent bits, with the nullability of unconstrained type parameter references initially set to `Nullable | NotNullable`? Then flow analysis might infer the reference is `NotNullable` only.

Reply:
With this structure, we should be able to accurately reflect what modifiers are in source code. Therefore, always pretending that there is a modifier when there isn't one in source, or always pretending that something is nullable when it isn't marked as nullable in source has a negative impact on checks we do on declarations. Checks on declarations should be performed specifically based on presence, or lack of '?' question mark in source code. Then, flow analysis can take that information coming from declarations and interpret it as it is appropriate for the purpose of flow analysis, however the structure itself shouldn't do this interpretation as it used to do before this change.  

---
In reply to: [230214729](https://github.com/dotnet/roslyn/pull/30913#discussion_r230214729) [](ancestors = 230214729)

Reply:
At some point I was thinking naming NotNullable as None (or something similar) to reflect exactly the fact that simply there was no annotation on the declaration, and then it is up to flow analysis to figure out what that means for a specific type. 

---
In reply to: [230218492](https://github.com/dotnet/roslyn/pull/30913#discussion_r230218492) [](ancestors = 230218492,230214729)